### ISSUE TITLE
Use keywords to define derivatives rather than hard-coding explicitly

### ISSUE THREAD

**Issue by leakec** [leakec/tfc#2]
## Feature
Rather than define each derivative transform we want individually for the basis functions, e.g., dH, d2H, etc., just use the already existent derivative value and a keyword. In the JVP transform, you can update increment the derivative value and then pass it to the original function.

This way we only need to define the derivative transform once (cleverly) and we get as many derivatives as we want for free! Furthermore, it will clean up the code significantly. Moreover, this should have no effect on the final JITed functions, as the XLA transforms already get this derivative value, i.e., its not adding any extra arguments in (honestly, even if it was, the derivative values are so small it probably wouldn't make any noticeable difference anyway).

A simple working example of this concept is shown below:
```python
import jax.numpy as np
from jax import core, abstract_arrays, jvp
from jax.interpreters import ad, batching, xla
from tfc.utils import egrad

x = np.array([1.,2.,3.])

f_p = core.Primitive("f")
def f(x, d=0):
    return f_p.bind(x, d=d)
def fImpl(x, d=0):
    return x+d
def f_jvp(arg_vals, arg_tans, d=0):
    x = arg_vals[0]
    dx = arg_tans[0]
    out_tans = f(x,d=d+1)*dx
    return (f(x,d=d), out_tans)
def f_abstract_eval(x, d=0):
    return abstract_arrays.ShapedArray(x.shape, x.dtype)
def f_batch(vec, batch, d=0):
    return f(*vec, d=d), batch[0]

f_p.def_impl(fImpl)
f_p.def_abstract_eval(f_abstract_eval)
batching.primitive_batchers[f_p] = f_batch
ad.primitive_jvps[f_p] = f_jvp

df = egrad(f)
d2f = egrad(df)
d3f = egrad(d2f)

print(d3f(x))
# For multivariate, make dinit = np.zeros((dim,))
# then just increment each piece based whether it has a derivtaive or not in jvp
```
In this example, we only define one JVP, but we get as many derivatives as we want (3 in this example). Obviously, this derivative transform is mathematically meaningless, its just for illustrative purposes. 

I think we should be able to do the same for multivariate, we just need to have the `d` keyword as an array, and increment the correct portions of the array as we go. The starting value (call it `d0`) can just be `np.zeros(dim)` where `dim` is the dimension of the problem specified to `mtfc`.

**Comment by leakec** [comment]
Closing as this was implemented by MR #3
